### PR TITLE
Backport catchall v62802

### PR DESCRIPTION
**Pull request by Axel-Naumann**

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
Will remove RISC-V commits,
```
* 35db349a73 Revert "FIXME: Comment asserts, they happen upstream but we believe they are harmless."
* 4c27a9cb06 D142443: [ORC] Drop Comdat when discarding IR symbol
* 502a58f946 D142352: [CodeGen] Filter out available_externally aliases
```
and `TChain: don't interpret anchor with parameters as treename`, on Monday.

File: interpreter/cling/lib/Interpreter/IncrementalJIT.cpp [hunk 1]
```diff
@@ -360,7 +474,32 @@ IncrementalJIT::IncrementalJIT(
...
+    // Default to JITLink on macOS, as done in LLVM by
+    // LLJITBuilderState::prepareForConstruction.
+    if (TT.isOSBinFormatMachO() &&
+        (TT.getArch() == Triple::aarch64 || TT.getArch() == Triple::x86_64)) {
+      jitLink = true;
+    }
+    // Finally, honor the user's choice by setting an environment variable.
+    if (const char* clingJitLink = std::getenv("CLING_JITLINK")) {
+      jitLink = cling::utils::ConvertEnvValueToBool(clingJitLink);
+    }
```
Comment (hahnjo) [inline interpreter/cling/lib/Interpreter/IncrementalJIT.cpp]: This commit should not be in here, we shouldn't add new "features" for patch releases

File: interpreter/cling/lib/Interpreter/IncrementalJIT.cpp [hunk 2]
```diff
@@ -459,8 +461,13 @@ CreateTargetMachine(const clang::CompilerInstance& CI, bool JITLink) {
...
     // LLJITBuilderState::prepareForConstruction.
-    JTMB->setRelocationModel(Reloc::PIC_);
-    JTMB->setCodeModel(CodeModel::Small);
+    JTMB.setRelocationModel(Reloc::PIC_);
+    // Set the small code except for macOS on AArch64 - it results in relocation
+    // targets that are out-of-range.
+    // TODO: Investigate / report upstream and re-evaluate after a future LLVM
+    // upgrade.
+    if (!(TT.isOSBinFormatMachO() && TT.getArch() == Triple::aarch64))
+      JTMB.setCodeModel(CodeModel::Small);
```
Comment (hahnjo) [inline interpreter/cling/lib/Interpreter/IncrementalJIT.cpp:460]: same, commit should not be here

File: interpreter/cling/lib/Interpreter/IncrementalJIT.cpp [hunk 3]
```diff
@@ -478,7 +478,7 @@ IncrementalJIT::IncrementalJIT(
   ErrorAsOutParameter _(&Err);
 
   LLJITBuilder Builder;
-  Builder.setDataLayout(TM->createDataLayout());
+  Builder.setDataLayout(m_TM->createDataLayout());
```
Comment (hahnjo) [inline interpreter/cling/lib/Interpreter/IncrementalJIT.cpp:483]: this commit was already backported in https://github.com/root-project/root/pull/12337!

**Comment by pcanal** [comment]
You will need: https://github.com/root-project/roottest/pull/935

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on ROOT-performance-centos8-multicore/cxx17

**Comment by Axel-Naumann** [comment]
The last build just verified that https://github.com/root-project/roottest/pull/941 fixes roottest.

**Pull request merged**